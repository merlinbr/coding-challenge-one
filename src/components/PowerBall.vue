<script setup lang="ts">
import { useQuery } from "@vue/apollo-composable"
import { computed, watchEffect } from "vue"
import type { IDrawHistory } from "@/interfaces/i-draw-history"
import { getDrawHistory } from "@/graphql/getDrawHistory"

const { result, loading } = useQuery(getDrawHistory, {
  filter: { lotteries: 'PowerBall' },
  licensedTerritory: 'US-NJ',
  limit: 5,
  offset: 0
})

const powerBallDraws = computed(() => result.value?.drawHistory.draws as IDrawHistory ?? [])
const powerBallLoadingMsg = 'Powerball Daten werden aktualisiert!' as string
const powerBallDateMsg = 'Die Powerball zahlen vom:' as string

function formatDate (date: Date): string {
  return new Date(date).toLocaleString('de-DE', {
    year: 'numeric',
    month: '2-digit',
    day: 'numeric',
  })
}
watchEffect(() => {
  console.log(powerBallDraws)
})
</script>

<template>
  <div class="container mx-auto p-4">
    <template v-if="loading">
      {{ powerBallLoadingMsg }}
    </template>
    <template v-else>
      <ul class="list-none text-white text-lg">
        <li
          class="mb-3"
          v-for="({ drawDate, drawnNumbers }, index) in powerBallDraws" :key="index"
          :data-test-id="'draw-date-' + index"
        >
          {{ powerBallDateMsg }} {{ formatDate(drawDate) }}
          <div
            v-for="number in drawnNumbers[0].numbers"
            class="space-x-2 text-white pt-2 text-lg"
            :key="number + index"
          >
            <template v-if="number.name === 'EXTRA'">
              <span class="mr-2">Extrazahl:</span>
            </template>
            <template v-if="number.name === 'POWERPLAY'">
              Powerplay:
            </template>
            <div
              v-for="value in number.values"
              class="bg-emerald-600 h-10 w-10 rounded-full text-center inline-block"
              :key="value"
            >
              <span class="align-center">
                {{ value }}
              </span>
            </div>
          </div>
        </li>
      </ul>
    </template>
  </div>
</template>
