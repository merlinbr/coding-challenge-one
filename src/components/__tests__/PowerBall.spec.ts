import { describe, it, expect, vitest } from 'vitest'

import { mount } from '@vue/test-utils'
import PowerBall from '../PowerBall.vue'
import { useQuery } from "@vue/apollo-composable"

vitest.mock('@vue/apollo-composable', () => ({
  __esModule: true,
  useQuery: vitest.fn(),
}))
useQuery.mockImplementation(() => ({
  result: { value: mockData }
}))

const mockData = {
  drawHistory: {
    draws: [
      {
        drawDate: '2023-03-26T02:59:00.000Z',
        addOns: ['powerPlay'],
        id: 'US_NJ_PowerBall_2023-03-26T02:59:00Z',
        drawnNumbers: [
          {
            numbers: [
              {
                name: 'BASE',
                values: [15, 17, 18, 47, 57]
              }
            ]
          }
        ]
      }
    ]
  }
}

describe('PowerBall', () => {
  it('renders properly', () => {
    const wrapper = mount(PowerBall)

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should show drawDate in specific date format dd.mm.yyyy', () => {
    const wrapper = mount(PowerBall)
    const drawDate = wrapper.find('[data-test-id="draw-date-0"]')

    expect(drawDate.text()).toContain('26.03.2023')
  })

  it('should show loading message when query is still loading', () => {
    useQuery.mockImplementation(() => ({
      result: {},
      loading: true
    }))
    const wrapper = mount(PowerBall)

    expect(wrapper.text()).toContain('Powerball Daten werden aktualisiert!')
    expect(wrapper.element).toMatchSnapshot()
  })
})
