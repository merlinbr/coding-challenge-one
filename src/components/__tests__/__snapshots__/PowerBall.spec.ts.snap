// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PowerBall > renders properly 1`] = `
<div
  class="container mx-auto p-4"
>
  <ul
    class="list-none text-white text-lg"
  >
    
    <li
      class="mb-3"
      data-test-id="draw-date-0"
    >
      Die Powerball zahlen vom: 26.03.2023 
      
      <div
        class="space-x-2 text-white pt-2 text-lg"
      >
        <!--v-if-->
        <!--v-if-->
        
        <div
          class="bg-emerald-600 h-10 w-10 rounded-full text-center inline-block"
        >
          <span
            class="align-center"
          >
            15
          </span>
        </div>
        <div
          class="bg-emerald-600 h-10 w-10 rounded-full text-center inline-block"
        >
          <span
            class="align-center"
          >
            17
          </span>
        </div>
        <div
          class="bg-emerald-600 h-10 w-10 rounded-full text-center inline-block"
        >
          <span
            class="align-center"
          >
            18
          </span>
        </div>
        <div
          class="bg-emerald-600 h-10 w-10 rounded-full text-center inline-block"
        >
          <span
            class="align-center"
          >
            47
          </span>
        </div>
        <div
          class="bg-emerald-600 h-10 w-10 rounded-full text-center inline-block"
        >
          <span
            class="align-center"
          >
            57
          </span>
        </div>
        
      </div>
      
    </li>
    
  </ul>
</div>
`;

exports[`PowerBall > should show loading message when query is still loading 1`] = `
<div
  class="container mx-auto p-4"
>
  
  Powerball Daten werden aktualisiert!
  
</div>
`;
